metricbeat/module/kubernetes/util/kubernetes.go

```diff
@@ -1142,11 +1086,8 @@ func (e *enricher) Stop(resourceWatchers *Watchers) {
 // This method is executed whenever a new event is created and about to be published.
 // The enricher's index method is used to retrieve the resource identifier from each event.
 func (e *enricher) Enrich(events []mapstr.M) {
-	e.RLock()
-	defer e.RUnlock()
-
 	for _, event := range events {
-		if meta := e.metadata[e.index(event)]; meta != nil {
+		if meta := e.getMetadata(event); meta != nil {
```

Comment:
with the lock moved from above just in the scope of `getMetadata` doesn't this mean that all of the below

```
k8s, err := meta.GetValue("kubernetes")
```

```
if podMeta, ok := k8sMeta["pod"].(mapstr.M); ok {
  event.DeepUpdate(podMeta)
}
```

are prone to race conditions?

Reply:
Good catch. I think in practice everything is fine, as we never actually modify the metadata after creating it. But I would like to rewrite this so the safety is immediately apparent. I think the simplest way of achieving this would be to make a copy in `getMetadata` and then avoid making copies here in `Enrich`. Let me try that.

Reply:
ok since we may do the Clone, let's be "eager" about it